### ISSUE TITLE
Incremental compilation should be supported with JDK 15

### ISSUE THREAD

**Issue by melix** [gradle/gradle#14744]
## Incremental compilation fails with JDK 15

Given the attached project, aimed at testing the behavior of the Java incremental compiler when we update a record component annotation, I noticed that we recompile everything if we change the annotation type.

This is probably because the incremental compiler doesn't support records, or doesn't support the JDK 15 class format.

Probably requires latest ASM.
[records.zip](https://github.com/gradle/gradle/files/5324161/records.zip)

**Comment by arash-carta** [comment]
Are there any updates on this issue?

We've started seeing this when running `compileJava` tasks with `--debug` flag, which definitely seems related:

```
org.gradle.api.internal.tasks.compile.incremental.classpath.DefaultClasspathEntrySnapshotter] Could not analyze XYZ.class for incremental compilation
java.lang.UnsupportedOperationException: This feature requires ASM8_EXPERIMENTAL
        at org.objectweb.asm.ClassVisitor.visitRecordComponentExperimental(ClassVisitor.java:304)
        at org.objectweb.asm.ClassReader.readRecordComponent(ClassReader.java:929)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:704)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:401)
        at org.gradle.api.internal.tasks.compile.incremental.asm.ClassDependenciesVisitor.analyze(ClassDependenciesVisitor.java:63)
        at org.gradle.api.internal.tasks.compile.incremental.analyzer.DefaultClassDependenciesAnalyzer.getClassAnalysis(DefaultClassDependenciesAnalyzer.java:41)
...
```

This happens upon any smallest change in any source file (record or not), and actually slows down compilation as it seems it makes incremental compilation to completely stop working. Just by eliminating few `record`s that we have, the incremental compilation comes back and compilation after small changes becomes fast again.

Fixing this probably requires Gradle to use ASM 8+.

Environment:

```
------------------------------------------------------------
Gradle 6.8
------------------------------------------------------------

Build time:   2021-01-08 16:38:46 UTC
Revision:     b7e82460c5373e194fb478a998c4fcfe7da53a7e

Kotlin:       1.4.20
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          14.0.2 (Oracle Corporation 14.0.2+12-46)
OS:           Mac OS X 10.15.7 x86_64
```

**Comment by melix** [comment]
@rieske would be good to check if this is fixed with the ASM upgrade in 7.0
